### DISCUSSION TITLE
Support for metadata

### DISCUSSION THREAD

**Discussion by petern3** [no-chris/chord-mark#532] [Ideas]
In the ChordPro format, the file format has support for lots of [metadata](https://www.chordpro.org/chordpro/chordpro-directives/) that is useful for:

* Displaying (e.g. key, capo, time signature, title, copyright)
* Sorting in a database (e.g. title, author, album, year)
* Integration by fancy software to play the music (e.g. tempo, duration)

I'm not suggesting the ChordMark format copies ChordPro or supports all of the metadata, but I think it would be really helpful to have some of this data as part of the format. These are the attributes I reckon are pretty fundamental:

* title
* key
* time signature
* capo
* custom tags

And these would be nice

* artist
* album
* composer
* copyright
* year

**Comment by no-chris** [comment]
Hello!

Thanks for your comment.

Metadata is definitely a must-have for interoperability yes.

I haven't prioritized it yet since there is no support for ChordMark in any 3rd party software (yet). Obviously, that can change at any time if there is interest in it, and that can be easily added to the core parsing library. Having application support in Chord Chart Studio will take a bit more time, though.

The idea was to implement metadata in Frontmatter (https://jekyllrb.com/docs/front-matter/) format. As for the field themselves, I agree with your suggestions.

Do you have a specific use case that needs solving quickly?

**Reply by petern3** [reply]
**tl;dr** The format is fine for me (for now) with no metadata support

You say that no third party software uses ChordMark. Unfortunately it's a bit of a catch-22 whereby the format won't get any support unless it's popular, but won't get popular unless people know about it. And missing metadata won't help either case.  I personally only came across it because I was considering resurrecting a [half-done](https://github.com/petern3/guitar_music_converter) project from a few years back so was doing a review of what's come up since then.

I have to say your project is exactly what I was after, and I'm particularly impressed with:

* The fact that you didn't go with the norm (chordpro) - I normally wouldn't approve with that decision, except that you've executed well
* That you've made an editor, complete with a renderer and print preview
* The reduction in how much source code is needed
* How you render the bars into the output format, it's very intuitive!

In other words, you've got what I think is the best product out there! (for my use case, basically the same as yours I think). That's off-topic, but I wanted to thank you somewhere.

Anyway, regarding metadata:

* Thanks for a format suggestion, I'll make use of that (and be willing to change it later to conform to whatever you settle on)
* Because it renders plain test, that's good enough for now
* Therefore I don't have a specific use case that needs solving quickly

**Reply by no-chris** [reply]
Thanks for the kind words, that is really encouraging 😁 

Nice project of yours, I can definitely see how that would be useful, especially as a building block for applications. Actually, my initial plan was not to write the converter but to rely on a third party lib, [ChordSheetJs](https://github.com/martijnversluis/ChordSheetJS). I even wrote the ChordSheetJs internal format adapter for Chord Chart Studio until I realised that the license was not compatible (GPL for ChordSheetJS, closed source - for now - for Chord Chart Studio). So I scratched everything and wrote the ChordMark converters from scratch 😅 

Do I understand correctly that you intend to use ChordMark somehow in your project?

**Reply by petern3** [reply]
What you've got is enough for me not to bother at all with my project. As useful as it could be for other people, it's only useful if it's actually adopted (which is unlikely) and if I maintain it (which I won't if I don't need to). Therefore I'm happy to jump completely on your project and work from that if I actually want to do anything significant!

The only thing I _may_ consider doing is creating a VS Code wrapper for it, since I like doing things locally
